programs/survival/waystones.sc

```diff
@@ -0,0 +1,310 @@
+);
+
+__on_player_dies(p) -> (
+    if(global_settings:'clear_waypoints_on_death', 
+        uuid = p~'uuid';
+        _read_player_waypoints(uuid);
+        delete(global_waypoints:uuid);
+        _write_player_waypoints(uuid);
+    ); 
+);
+
+_handle_name_tag(item, pos, p, hand) -> (
+    if(item:0=='name_tag' && item:2:'display',
+        global_waystones:pos:'icon':2 = item:2;
+        global_waystones:pos:'name' = _get_name_from_nbt(item:2);
+        slot = if(hand=='mainhand', p~'selected_slot', -1);
+        inventory_set(p, slot, item:1 - 1);
+        sound( 'block.enchantment_table.use', pos );
+        _app_message(p, 'Waypoint named '+global_waystones:pos:'name');
+    true, false);
+);
+
+_mark_player_waypoint(stone_pos, point_pos, uuid) -> (
+    if(!global_waypoints:uuid, global_waypoints:uuid = {}); 
+    global_waypoints:uuid:stone_pos = map(point_pos, floor(_));
+);
+
+// op's waystone screen
+open_waystones_screen() -> (
+    p = player();
+    icons = map(pairs(global_waystones), [_:0, _:1, _:0 + [1,0,0]] );
+    if(length(icons) > 0,  
+        _print_icons_to_screen(_create_warps_screen(p,'fb All Waystones'), icons);
+    ,
+        _app_message(p, 'No waystones to display');
+    );
+);
+
+// open when player clicks on a lodestone
+_open_waypoint_screen(p, pos, uuid) -> (
+    icons = [];
+    bad_keys = [];
+
+    // merge waypoints and waystones data into icons.
+    for(pairs(global_waypoints:uuid),
+        if( global_waystones:(_:0), 
+            icons += ([_:0,global_waystones:(_:0),_:1]);
+        , 
+            bad_keys += _:0;
+        );
+    );
+
+    // Clear player waypoints with missing waystone entires
+    for(bad_keys, delete(global_waypoints:uuid:_));
+
+    // filter out entires in different dimensions.
+    if(!global_settings:'dimensional_crossing', 
+        dimension = current_dimension();
+        icons = filter(icons, _:1:'dimension' == dimension);
+    );
+
+    if(length(icons) > 0,  
+        _print_icons_to_screen(_create_warps_screen(p, str('fb Visited Waypoints', p~'name')), icons);
+    );
+);
+
+_create_warps_screen(p, title) -> (
+    create_screen(p, 'generic_9x6', format(title), _(screen, p, action, data) -> (
+        if(data:'slot' > 53, return());
+        if(action=='pickup',
+            btn = inventory_get(screen, data:'slot');
+            if(!btn, return('cancel'));
+            btn = parse_nbt(btn:2);
+            _warp_player(p, btn:'pos', btn:'dimension', btn:'point');
+            close_screen(screen);
+        );
+        return('cancel');
+    ));
+);
+
+_print_icons_to_screen(screen, icons) -> (
+    for(slice(sort_key(icons, _:1:'name'),0, min(length(icons),53)),
+        stone = _:1;
+        nbt = nbt({
+            'pos'->_:2,
+            'point'->_:0,
+            'dimension'->stone:'dimension' || 'overworld',
+            'display'->{
+                'Lore'->[escape_nbt(str('"%d, %d, %d"', _:0))],
+                'Name'->escape_nbt( str('"%s"', stone:'name'))
```

Comment:
I think it'd be cool to have some formatting applied to the raw data. Right now them getting italic or the amazing vanilla purple lore isn't that great IMO, and this is basically the most visible part of the app.

Reply:
I am trying this, and as far as i can tell formatting the item name or lore with the `format` scarpt method is not working. It looks possible in default mc but it seems slightly more complicated then I want to think about rn.